src/librustc_trans/trans/attributes.rs

```diff
@@ -265,7 +265,7 @@ pub fn from_fn_type<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>, fn_type: ty::Ty<'tcx
                 // on memory dependencies rather than pointer equality
                 let interior_unsafe = mt.ty.type_contents(ccx.tcx()).interior_unsafe();
 
-                if mt.mutbl == hir::MutMutable || !interior_unsafe {
+                if mt.mutbl != hir::MutMutable && !interior_unsafe {
```

Comment:
Yes, and that's intentional. The change is supposed to only affect `&mut` references (and '&' ones with unsafe interiors), because those are the ones that could have stores that would expose this bug.